### ISSUE TITLE
execute_convergence should have success, try again, failure semantics

### ISSUE THREAD

**Issue by lvh** [rackerlabs/otter#956]
Currently, `execute_convergence` has return type `Effect bool`. The bool tells you if you should retry or not, and it picks that based on whether it executed any steps this cycle.

There's three reasons you might want to converge again or not converge anymore:

1. Convergence is done. (Stop converging.)
2. Convergence isn't going to happen; the group is wedged. (Stop converging.)
3. Convergence is happening, but we're not done yet. (Keep converging.)

Right now it returns `True` if some steps/effects were executed, `False` otherwise. However, this means that if some steps were executed, but the group is broken, it'll return `True`, and the caller will retry. That means we'll spend however many convergence cycles we're normally willing to wait for a non-broken group to converge for that broken group to converge, which we know is never going to happen.

If wedged, we should also set the group status to `ERROR`.

We can either add "figure out if wedged" behavior to `execute_convergence`, or just return `Effect [results]` instead of `Effect bool` and have some other function do it. This probably means changing all of the steps' effects. A step knows when it has failed, and it's not reasonable for that callback to know about the failure modes of all steps ever. Therefore, the return types of step effects should have something along the lines of:

```python
from twisted.python.constants import ValueConstant, Values
class Status(Values):
    """
    Constants representing success types.
    """
    OK = ValueConstant(1)
    RETRY = ValueConstant(2)
    FAIL = ValueConstant(3)
```

(These are value constants just so I can find the worst case in a group with `max`. It's fine if they're named instead.)

Suggestions:

- If a step's effect (partially) fails, it's allowed to optimistically execute some more effects. It's of type `Effect Status`, i.e., it eventually gives you a status.
- Step effects are of type `Effect (Status, [Effect])`. If none of its sibling effects (i.e. other effects passed to the same `parallel` call as part of an `execute_convergence` attempt) declare the group to be wedged, these effects are executed instead. This has the benefit of allowing us to encode graceful recovery of error conditions that we can figure out without another entire gathering cycle.

The return type of `execute_convergence` could stay `Effect bool`, or become `Effect Status`; I'm not sure it matters.

**Comment by radix** [comment]
For the sake of factoring, I would suggest `execute_convergence` return `Effect Status`, which separates "how to do convergence" from "when to do convergence". Then, the "when to do convergence" layer will be responsible for marking the group as wedged when it sees a Status of `FAIL`.

I understand what you meant about the additional list of Effects now -- describing them as "optimistic" cleanup effects helped. Could you give an example of what these effects might do for a particular step? Does it really need to be a list?

**Comment by lvh** [comment]
Doesn't have to be a list, it could be a single `Parallel` effect just as well :)

Concrete example: consider the following RCv3 (LB, node) pairs:

```
[
        ("lb1", "node1”),
        ("lb2", "node2”)
]
```

(Imagine for a second those are UUIDs.)

If I issue a bulk delete for the following two pairs:

```
[
        ("lb1", "node1”),
        ("lb2", “some-old-node”)
]
```

I'm going to get a 409 Conflict response:

```
{
    "errors": [
        "Node some-old-node is not a member of Load Balancer Pool lb2"
     ]
}
```

RCv3 bulk deletes are atomic, so the `("lb1", "node1")` pair will *not* have been deleted. While we could just reconverge (and hopefully RCv3 is consistent, so we won't decide to try and get rid of the `some-old-node` pair again), we could also just issue another delete for the remaining pair.

**Comment by lvh** [comment]
On a more general note, this seems to be a pattern that comes back: do we just reconverge as our only error recovery mechanism? Originally, steps weren't going to have any retry logic at all, but we've come back from that; I think this is just another way of doing something slightly smarter with hopefully as little ugliness as possible.

**Comment by lvh** [comment]
Also, if all of the steps have `Status.OK`, we can do away with an extra convergence cycle (which would do nothing, so just the gathering step) I guess?

**Comment by radix** [comment]
@lvh given the way we do throttling I think we still need the toplevel thing to return `RETRY` until there are no steps

**Comment by radix** [comment]
So, @lvh, I don't understand why the RCv3 node deletion thing needs this API. Why can't it just retry immediately upon failure that type of failure? Why does it need to wait for all the other steps to be done?

**Comment by lvh** [comment]
Re: toplevel thing: isn't that what `ConvergeLater` (#818) is for? We can't do without, because it's possible that you're just waiting for machines to build, but you already have the correct number of machines being built. Results in empty plan (right now), but you're clearly not done. (We could also count on AtomHopper to notify us to retry converging at a later point, but then Otter functioning correctly hinges on this other thing over here functioning correctly.)

Re: API: assuming you're referring to `Effect (Status, Effect)`: nope, it doesn't actually need that. The two suggestions are mutually exclusive. The `Effect Status` one works too, with the caveat that it might do some pointless work for wedged groups.

**Comment by lvh** [comment]
Oh, and I guess it's `Effect (Status, String)`, because you want to be able to report reasons?

Anyway, `ConvergeLater` would just be a `Constant` of `(Status.RETRY, "Not done yet, waiting on some events to complete")`. Potentially that message should be part of the state of `ConvergeLater`, so that the thing that creates that `ConvergeLater` instance (who knows why you want to converge later) can document the reason.

**Comment by radix** [comment]
Okay, then I personally suggest we leave out the list of effects until we find a more compelling use case for them.

I'm still kind of confused about the whole ConvergeLater thing -- I guess I need to more thoroughly read the discussion that's happened around it.

**Comment by radix** [comment]
Alright, so we had a big conversation about this on Slack, especially in how it relates to the "`ConvergeAgain`" idea that has been passed around for a while.

`ConvergeAgain` should only be a step if it can be treated like any other step, and doesn't need to be special cased by any other code.

Assuming a Step.as_effect signature of `Step -> (Status, reason)` Its `as_effect` method would return Effect(Constant((Status.RETRY, "need to retry")))`. However, we may need to extend this with timing information. We have many different cases in which we want to reconverge:
- we're not executing all the steps we need to (e.g., we've throttled some CreateServers)
- we're waiting for some amount of time to pass before we can perform some other operation (e.g. waiting for a node to drain from the LB for 5 minutes before we delete it)
- an ephemeral error has occured and we want to retry convergence.

The first two of these cases are use cases for `plan()` returning a `ConvergeAgain` step. The last one is a use case for other steps being able to also indicate the desire to converge again (e.g. the RCv3 node deletion use case from earlier in this ticket).

So, to get back to timing: if these things have significantly different timeouts (perhaps 0 seconds for an ephemeral-error-retry, and 5 minutes for a LB node draining), it's probably useful to add a time at which we want to reconverge.

But, what is the actual range of times we want to reconverge on? If 5 minutes is too high of an example, and most times are between 0 seconds and 30 seconds, it may be sufficient to just leave that out of the interface (for now) and always use something like 30 seconds. I think we've factored things well enough that adding it later should be easy.

**Comment by lvh** [comment]
> Assuming a Step.as_effect signature of `Step -> (Status, reason)`

This is `Step -> Effect (Status, reason)`, right?

**Comment by radix** [comment]
@lvh oh yes, sorry for the mistake. I'll edit.

**Comment by lvh** [comment]
How does this integrate with all of our `success_pred` stuff? That wants to raise-on-error, which we can change; but where does the reason come from?

**Comment by lvh** [comment]
I guess maybe I should be asking: why isn't `success_pred` something attached in `IStep.as_effect` like this:

```
return service_request(whatever).on(my_first_success_predicate)
```

?

**Comment by radix** [comment]
It could very well just be a callback that you attach to the effect. In fact, many of the features that service_request does could have an interface like that; they're only implemented in service_request for convenience and to allow for default behavior.

**Comment by lvh** [comment]
What I'm (perhaps too opaquely) alluding to is that if we're going to start returning a status *and* a reason, it's a lot trickier to generalize these functions. I guess we could just change `has_code` to work somewhat like:

`has_code({200: (Status.OK, ""), 409: (Status.RETRY, ""), 500: (Status.RETRY, ""), 418: (Status.FAIL, "server is actually a teapot")})`

... and even then, at that point it doesn't seem to provide anything over the more generic `.on` interface, right?

**Comment by lvh** [comment]
Okay, so to recap: next, we're going to start turning all of the `IStep.as_effect` effects into ones that are of type `Effect (Status, reason)`, right?

**Comment by lvh** [comment]
I think it would be a good idea to leverage the List monad to encapsulate the nondeterminism of the number of reasons, i.e. make it `Effect (Status, [Reason])`. Advantages:

1. "No reason" looks a lot better as the empty list (explicitly: no reasons) rather than the empty string.
2. Things that collect reasons only have to know about plain old concatenation. This makes it easier to have better objects than strings to describe what the reason actually is.

**Comment by radix** [comment]
@lvh makes sense. +1

**Comment by cyli** [comment]
We have encountered various issues with the worker using the Bulk RCv3 steps - if the return value of an effect is (Status, reason), then we can't use the Bulk steps for adding, since the worker relies on having the output of adding to a load balancer.  

This is not to say that we should not make this change, just that something needs to be done for the worker in the meantime.  See discussion in #1030.

**Comment by radix** [comment]
I think the best way to move forward would be to just factor out the common code in a way that's usable by _rcv3.py and the steps. Something like bulk_add_to_rcv3(lb_node_pairs) -> Effect (treq_response, body) which is called by both _rcv3.py and the Step code. The Step code would further translate that to a Step Status, but the _rcv3.py code would use the body directly.

**Comment by radix** [comment]
This was fixed by #1219.
